app/scripts/lib/router.js

```diff
@@ -134,6 +136,27 @@ define(function (require, exports, module) {
       this.storage = Storage.factory('sessionStorage', this.window);
     },
 
+    onSignUp () {
+      const View = this._isEmailFirstFlow() ? SignUpPasswordView : SignUpView;
+      return this.showView(View);
+    },
+
+    onSignIn () {
+      const View = this._isEmailFirstFlow() ? SignInPasswordView : SignInView;
+      return this.showView(View);
+    },
+
+    /**
+     * Is the user in the email-first flow?
+     *
+     * @returns {Boolean}
+     * @private
+     */
+    _isEmailFirstFlow () {
+      const viewModel = this.getCurrentViewModel();
```

Comment:
This makes the assumption that the view the user came from set the `emailFirst` field on the model. This assumption doesn't always hold, I got into a strange state today going to either /reset_password or /choose_what_to_sync, I'm not sure which one - one of them sent me back to the `/signup` page w/o specifying `emailFirst` and I had the old flow.

Once in the `emailFirst` flow, always in the emailFirst flow.